File: Python/MinimumIndexSumOfTwoLists/MinimumIndexSumOfTwoLists.py
class Solution(object):
	# (122ms)
    def findRestaurant(self, list1, list2):
        """
        :type list1: List[str]
        :type list2: List[str]
        :rtype: List[str]
        """
        dt, ans, minSum = {}, [], len(list1) + len(list2)
        for index, s in enumerate(list1):
        	dt[s] = index
        for index, s in enumerate(list2):
        	if s in dt:
        		if index + dt[s] > minSum:
        			continue
        		if index + dt[s] < minSum:
        			ans = [s]
        			minSum = index + dt[s]
        		else:
        			ans.append(s)
        return ans

print Solution().findRestaurant(["Shogun", "Tapioca Express", "Burger King", "KFC"], 
	["Piatti", "The Grill at Torrey Pines", "Hungry Hunter Steakhouse", "Shogun"])

File: JavaScript/MinimumIndexSumOfTwoLists/MinimumIndexSumOfTwoLists.js
/**
 * (212ms)
 * @param {string[]} list1
 * @param {string[]} list2
 * @return {string[]}
 */
var findRestaurant = function(list1, list2) {
    var map = new Map();
    for(var i = 0; i < list1.length; i++) {
    	map.set(list1[i], i);
    }
    var ans = [], minSum = list1.length + list2.length;
    for(i = 0; i < list2.length; i++) {
    	if(map.has(list2[i])) {
    		if(i + map.get(list2[i]) < minSum) {
    			ans = [];
    			ans.push(list2[i]);
    			minSum = i + map.get(list2[i]);
    		} else if(i + map.get(list2[i]) == minSum) {
    			ans.push(list2[i]);
    		}
    	}
    }
    return ans;
};